src/user/user.service.ts

```diff
@@ -0,0 +1,58 @@
+import { Injectable } from '@nestjs/common'
+
+import { ConflictError } from '../common/errors/errors'
+import { hash } from '../common/utils/crypto.utils'
+import { CreateUserDto } from './dto/create-user.dto'
+import { UpdateUserDto } from './dto/update-user.dto'
+import { UserDatabaseService } from './user.database.service'
+import type { User } from './schemas/user.schema'
+
+// TODO: return all user data without passwords?
+
+@Injectable()
+export class UserService {
+  constructor(private readonly userDatabaseService: UserDatabaseService) {}
+
+  async isUserExist(username: string): Promise<boolean> {
+    return await this.userDatabaseService.isUserExist(username)
```

Comment:
As I commented earlier, it's a business logic so should be _placed_ here.
From the DB layer perspective, you already have all needed methods (`.userDatabaseService.findUserByUsername` for example) — utilize it here to decide if user exists or not.

Moreover, this seems redundant as well (at this point at least), since this method is not used outside of the UserService.
Placing duplication check inside `UserService.createUser()` looks more straightforward.

Reply:
Thinking further, we should ask ourselves if it's even needed at all. As far as I remember, `username` uniqueness is a DB-level constraint (you've defined unique index for `username`).
With this in mind, you'll not be able to create multiple records with the same `username`; so, I assume, `this.userDatabaseService.createUser` will throw some native error in such case. Can't we handle this error?

Reply:
> As I commented earlier, it's a business logic so should be _placed_ here. From the DB layer perspective, you already have all needed methods (`.userDatabaseService.findUserByUsername` for example) — utilize it here to decide if user exists or not.

I can't use it in this way because of error handling inside the `findUserByUsername`. 
```
async findUserByUsername(username: string): Promise<User> {
    const user = await this.userModel.findOne({ username })

    if (!user) {
      throw new NotFoundError(`User with username ${username} not found`)
    }

    return user.toObject()
  }
```

Reply:
> Thinking further, we should ask ourselves if it's even needed at all. As far as I remember, `username` uniqueness is a DB-level constraint (you've defined unique index for `username`). With this in mind, you'll not be able to create multiple records with the same `username`; so, I assume, `this.userDatabaseService.createUser` will throw some native error in such case. Can't we handle this error?

Yeah, I thought about this before. But it's not handles for now with a detailed error:
```
{
    "message": "Internal Server Error",
    "statusCode": 500
}
```

Reply:
I'm talking about the error thrown by `mongoose` when you're trying to create document that violates index constraints. Is it any descriptive so that we can handle it in a special way?
The error you've referenced above is obviously not a Mongo error.